[12.0] base_rest: New api

lmignon opened the pull request:
fixes #6 and #2 
includes #41

lmignon committed:
[ADD] datamodel: New addon allowing you to define dataclass like models serialisable to/from json/yaml

lmignon committed:
[IMP] base_rest: New API

lmignon committed:
[IMP] datamodel: registry is always available a class level
We no more need to get the registry from the context since it's always available on the schema class

lmignon committed:
[FIX] datamodel: schema property of NestedModel must be a class
This change fixes the incompatibility with the latest marshmallow version

lmignon committed:
[IMP] datamodels: ensures that the Odoo environment is always available as a property of datamodel instances as well as on marshmallow schema instances.

lmignon committed:
[IMP] datamodel: marshmallow 2.0 is released

lmignon committed:
[FIX] base_rest: lookup for service components registed for the given collection
This specific lookup is required to filter out component without collection. In the same time we must also filter out components that are not services

lmignon committed:
[FIX] base_rest: Avoid property resolution when inspecting object members
This is required to avoid some trouble with properties resolved by the method inspect.getmembers and accessing resources not yet completly initialized

kevinkhao committed:
Minor dependency on datamodel fix

kevinkhao committed:
Minor typo fixes

kevinkhao committed:
[IMP] Add decimal to float conversion for JSON serializing

kevinkhao committed:
[IMP] Return explanatory message for invalid JSON request instead of server error if a json request is made with bad json

lmignon committed:
Merge pull request #2 from akretion/12.0-new-api-wip-dependencies

Minor dependency on datamodel fix

lmignon committed:
Merge pull request #3 from akretion/12.0-new-api-wip-typos

Misc minor improvements: typo fixes, add Decimal JSON serializer

lmignon committed:
Merge pull request #4 from akretion/12.0-new-api-wip-json

12.0 new api wip json

sbidoul committed:
Use the latest pylint-odoo

sbidoul committed:
Replace missing-manifest-dependency check

This check does not work when we use PyPI names
in the manifest. Such an error will be caught
by tests when we install dependencies with pip.

sebastienbeau commented on base_rest/models/rest_service_registration.py:
```diff
@@ -49,6 +58,52 @@ def _register_hook(self):
             (name, cls) for name, cls in controllers if "RestController" not in name
         ]
         http.controllers_per_module["base_rest"] = controllers
+        # create the final controller providing the http routes for
```
I am testing this new version of rest api, and it's seem that we have some issue here.
If I stop/start odoo and refresh quickly odoo in my web browser it fail to load the registry correctly and my server is crashed for ever, so I need to start and wait some second.
Also If I run the base_rest_demo/test/test_openapi.py I have the issue with the registry. I am not sure that it's related.

@lmignon can you try in local to see if you have the same issue?

sebastienbeau replied:
It's seem that I do not have the issue anymore.

sbidoul committed:
Pin graphene<3

Graphene >= 3 dependes on graphql-core > 3,
which is incompatible with graphql-server-core.

nilshamerlinck committed:
[FIX] http/json error on first json request of new worker

lmignon committed:
Merge remote-tracking branch 'oca/12.0' into 12.0-new-api-wip

lmignon committed:
[FIX] base_rest: Use apispec>=4.0.0

etobella commented on base_rest/restapi.py:
```diff
@@ -0,0 +1,250 @@
+        elif isinstance(self._schema, str):
+            # schema is a method name to call on service to get the schema or
+            schema = getattr(service, self._schema)()
+        if isinstance(schema, Validator):
```
I think this needs to be 
```
if isinstance(self._schema, Validator):
    return self._schema
```

etobella reviewed (changes requested):
Great Work. Tested and it works perfectly :wink:
I found a small error

etobella commented on base_rest/apispec/base_rest_service_apispec.py:
```diff
@@ -0,0 +1,74 @@
+                self.path(
+                    path,
+                    description=description,
+                    operations={method.lower(): {}},
```
Change it for:
```
                self.path(
                    path,
                    operations={method.lower(): {'summary': description}},
```
Right now, the description is not shown. With this, it is visible :wink:

lmignon replied:
Thank you I've to check

etobella commented on base_rest/restapi.py:
```diff
@@ -0,0 +1,250 @@
+        try:
+            return ModelClass.load(
+                params, many=self._is_list, unknown=marshmallow.EXCLUDE
+            )
```
a `.dump()` is required here. It is not working when input json contains a list with a dict inside.

lmignon replied:
`dump()`???? I don't follow you. If your input is a list, you must declare your param's method as `Datamodel('my.rest.model', is_list=1)`

etobella replied:
I know, but it was not received properly on the rest function.

etobella replied:
My error defining the function. Sorry

etobella commented:
@lmignon I am testing your improvement and it is awesome. Much more easier :wink: I am adding the comments with the problems I detected, or do you prefer a PR?

etobella commented on base_rest/restapi.py:
```diff
@@ -0,0 +1,250 @@
+
+    def to_response(self, service, result):
+        validator = self.get_cerberus_validator(service)
+        if validator.validate(result):
```
Not covered what will happen if it is a list. I have tested with:
```
    def to_response(self, service, result):
        validator = self.get_cerberus_validator(service)
        if self._is_list:
            data = []
            for res in result:
                if not validator.validate(res):
                    raise SystemError(_("Invalid Response %s") % validator.errors)
                data.append(validator.document)
            return data
        if validator.validate(result):
            return validator.document
        raise SystemError(_("Invalid Response %s") % validator.errors)
```
And worked perfectly

lmignon replied:
@etobella Are you sure you use the right version of marshmallow_object and marshamallow? AFAIK is_list is natively supported by the marshmallow_object.... I've to check since I'm pretty sure that should work as is..

etobella replied:
It didn't work on my side. I will check again just to be sure

etobella commented on base_rest/restapi.py:
```diff
@@ -0,0 +1,250 @@
+
+    def from_params(self, service, params):
+        validator = self.get_cerberus_validator(service)
+        if validator.validate(params):
```
Not solved what will happen with self._is_list

lmignon commented on base_rest/restapi.py:
```diff
@@ -0,0 +1,250 @@
+        return {"200": {"content": {"application/json": {"schema": json_schema}}}}
+
+    def get_cerberus_validator(self, service):
+        schema = None
+        if isinstance(self._schema, dict):
+            # schema is a cerberus schema
+            schema = self._schema
+        elif isinstance(self._schema, str):
+            # schema is a method name to call on service to get the schema or
+            schema = getattr(service, self._schema)()
+        if isinstance(schema, Validator):
+            return schema
```
@etobella IMO the right fix should be:
```suggestion
        schema = self._schema
        if isinstance(self._schema, dict):
            # schema is a cerberus schema
            schema = self._schema
        elif isinstance(self._schema, str):
            # schema is a method name to call on service to get the schema or
            schema = getattr(service, self._schema)()
        if isinstance(schema, Validator):
            return schema
```
We must add a test to cover this case

etobella replied:
Good for me

lmignon commented:
@etobella Thank you for your reviews.... I'll try to take a look to your reported issues ASAP but I've a lot of work...

Regards

etobella commented:
If it is not a problem for you, we can make a PR or take the PR from your side (keeping your commits, obviously :smile: )

etobella committed:
[FIX] base_rest: finish new rest API

lmignon commented:
@etobella I'm not convinced by some of your proposed changes that looks to me like a workaround an other problem (the management of is_list). I would be happy to have this PR merged... but I want to be sure that there is no other to fix the issues you 've encountered. It would be nice to add the required tests to highlight these issues.

Regards,

lmi

etobella commented:
We can remove the part of is_list if you are not confortable with it

etobella commented:
According to Cerberus API Documentation (https://docs.python-cerberus.org/en/stable/api.html), the Validator must have an schema and no information about a List is provided on the first element, so you are unable to return `[{'a': 1}, {'a': 2}]`. You can make it with datamodels, not with CerberusValidator

That part was intended to solve this issue, at least on the response part (maybe we shouldn't implement it on the receive, I don't know). However, as I said, if you are not comfortable with it, we can remove it in order to solve it.

lmignon commented:
@etobella My bad, part of your proposals are for Cerberus not for datamodels. Nevertheless,  the one for the datamodel looks suspicous to me...
@sebastienbeau What's the status of your tests with this new implementation?

etobella commented:
@lmignon Don't worry, I was capable to solve most of my problems. At the end, the only changes necessary where this ones:

https://github.com/tegin/rest-framework/commit/545c4e1ee05db5212eb723d92751abc86fb72dba

lmignon commented:
@etobella You need to fix the tests since one of the changes replaces the 'summary' key into openapi document with 'description'
https://travis-ci.com/github/tegin/rest-framework/jobs/441823733#L859

sebastienbeau commented:
@lmignon I still have this weird issue of missing database in _component_databases[dbname]

This is random I reproduce it locally and also on travis see this job that have failed
https://travis-ci.com/github/shopinvader/odoo-shopinvader/jobs/428039949#L2026
Do you have the issue in local ? Try with the shopinvader PR

Note on shopinvader I have to change some code see my PR here : https://github.com/shopinvader/odoo-shopinvader/pull/802

I make more test soon.

etobella commented:
@lmignon All tests fixed in my branch, at the end, I made two commits, but they can be squashed :wink: 
https://github.com/tegin/rest-framework/tree/12.0-new-api-wip

lmignon commented:
@etobella Thank you for all your work. I'll merge your commit. I'll also split base_rest into 3 addons.
* base_rest (core functionalities)
* base_rest_cerberus (base_rest with support for cerberus message definitions)
* base_rest_datamodels (base_rest with support for datamodels message bindings)

@simahawk, @sebastienbeau  What do you thing about the split proposal?

lmignon committed:
[IMP] base_rest: Move code to support datamodel into a dedicated addon

lmignon committed:
[ADD] setup.py

lmignon committed:
[FIX] pre-commit: Fix pylint

lmignon committed:
[IMP] base_rest: Ensure controller unicity by service and collection

lmignon committed:
[IMP] base_rest, datamode: Spelling

lmignon committed:
[IMP] base_rest: Improve code readability

lmignon committed:
[IMP] datamodel: Move DataModelFactory class definition outside the datamodels method implementation

lmignon committed:
[IMP] base_rest: Improve documentation

lmignon committed:
[FIX] base_rest: Add missing http routes for service methods declared without the restapi.method decorator

lmignon committed:
[IMP] base_rest: Add tests for the openapi document generation process

lmignon committed:
[IMP] base_rest: Add tests for CerberusValidator

lmignon committed:
fix website into __manifest__.py

lmignon committed:
[IMP] base_rest: remove deprecated call to inspect.getargspec

lmignon commented:
/ocabot merge major
